### ISSUE TITLE
Alacritty on Wayland paste issue 

### ISSUE THREAD

**Issue by joshuachp** [alacritty/alacritty#2795]
# System
|                      |                                     |
|------------------|-------------------------------|
| Operating System | Arch Linux |
| Display Server   | Wayland |
| Window Manager   | Gnome |
| Alacritty | 0.3.3 |

# Summary
When you past some text in the terminal it respond with an error

# Behavior
Pasting some text causes an error depending on the shell:
* **zsh** doesn't paste the text and 
`bracketed-paste-magic:zle:47: not enough arguments for -U`
* **bash** doesn't past the text

I can't reproduce it 100% of the time but it happens often. With the same setup it doesn't happen in Tilix or the gnome terminal.

**Comment by kchibisov** [comment]
Would you mind testing the latest master?

**Comment by joshuachp** [comment]
The behavior is still the same in the one built from source.

**Comment by joshuachp** [comment]
> How are you pasting text? Ctrl+Shift+V or Middle Mouse?

With both methods the issue is present

> Would you mind testing the next scenario?
> 
> Launch alacritty from the latest master.
> Copy something in that alacritty instance.
> Paste right back to it.

I tried multiple times and I didn't get the error when I copied and pasted something from the same alacritty instance. Although it's  present if you paste across two different instances.

**Comment by ranisalt** [comment]
What's next? It's not very useful to copy and paste only inside a single instance of a software. I noticed it started after `wl-clipboard` was updated to 2.0 from 1.x, but I'm not sure it's related.

**Comment by kchibisov** [comment]
@ranisalt 

This is probably indirectly related. The main problem is that I can't repro this issue at all on systems I have. I wonder if clipboard works between `alacritty` instances though, since we kind of have this problem reported.

Are you also on GNOME btw?

**Comment by ranisalt** [comment]
Yep, I reported that because it might be the same issue affecting different packages. I'm using GNOME 3.34 and alacritty master.

**Comment by tones111** [comment]
I'm also unable to paste (ctrl+shft+v or mouse button binding).  I've verified that Alacritty can push text into the clipboard by pasting it into other applications.

  When stepping through the debugger it appears the problem lies in smithay-clipboard.  At threaded.rs:341 I see the ThreadRequest::LoadPrimary case being taken, however the
`primary_selection_device_manager.lock().unwrap().is_some()` returns none as does the attempt to get the gtk_primary_selection_device_manager so the content is being set to String::new().  Maybe smithay-clipboard should provide some sort of error indication.

This is on a system running KDE Plasma with Alacritty master (856cddc)

**Comment by webhive** [comment]
Confirm. Ctrl+shft+v does not work. Shift + Ins to paste selection does not work as well. 

```
alacritty --version
alacritty 0.3.3 (76dd84be)

GNOME 3.34.0
```

**Comment by kchibisov** [comment]
@webhive please update to the latest master or at least include Wayland clipboard fixes that we've shipped after that commit and retest it.

**Comment by HarlemSquirrel** [comment]
I just built [alacritty-git](https://aur.archlinux.org/packages/alacritty-git) 0.3.3.1315.g87cf14a4-1 and I can copy and paste on GNOME 3.34 on Wayland!

**Comment by kchibisov** [comment]
@HarlemSquirrel fyi your issue was addressed here #2574.

**Comment by bugaevc** [comment]
> I noticed it started after wl-clipboard was updated to 2.0 from 1.x, but I'm not sure it's related.

Just to confirm, this turned out to be completely unrelated to the wl-clipboard update, correct?

**Comment by kchibisov** [comment]
@bugaevc , We're not using `wl-clipboard`, so should be unrelated.

**Comment by kchibisov** [comment]
@tones111 I'm not sure if primary selection available on KDE at all, since QT doesn't support it on Wayland yet iirc. In case gtk_primary_selection and primary_selection are `None` that means, that your compositor doesn't support this protocol, so it's not an error strictly speaking. Would you mind logging just `Load` and `Store`, since they should work at least.

**Comment by tones111** [comment]
Changing the Primary ClipboardProvider calls to use load and store fixes the clipboard interaction for me.

**Comment by kchibisov** [comment]
Do you have `save_to_clipboard: true` in your config? Change it to false if this the case. Ctrl + Shift + V/C should work for you then.

**Comment by p4t0n1x** [comment]
Hi I'm having the same issue, I'm using the related software:

debian buster
alacritty 0.3.3
tmux 2.8
gnome 3.30.2

When having this setting to **false** "save_to_clipboard", ctrl + v and middle mouse button fails to paste the data copied.

If a set to **true** the setting "save_to_clipboard" ctrl + v works.

My alacritty instance was installed using the binary.

I hope this information can help.

**Comment by kchibisov** [comment]
@p4t0n1x your issue should be fixed on the latest master.

**Comment by leehambley** [comment]
I can't confirm your exact setup, but I was fortunate to get help from @pschyska IRL to help me build and install my own package from master, and it works flawlessly for me now. 

I recommend trying to build your own pacakge whilst waiting for the next release of alacritty.

**Comment by p4t0n1x** [comment]
Awesome, then I will try to build it by myself, thank you both.

P.D. Alacritty is amazing, thanks!!

**Comment by breuerfelix** [comment]
is there any guide on how to build alacritty from master ?

i try it with `cargo install --git <repo_url> --branch master`

but so far no luck.... i got some errors to fix, is there another way ?

**Comment by chrisduerr** [comment]
> its not linked in the readme so i was a little blind :)

It is: https://github.com/jwilm/alacritty/#installation :)

**Comment by smason1995** [comment]
I'm having this issue on openSUSE Tumbleweed with 
TMUX 3.0a
Alacritty 0.3.3
Gnome 3.34.1

In my case, it does not matter if my `save_to_clipboard` is equal to true or false (tried both, still can't paste). I'm using the keyboard shortcut, haven't tried mouse cause I am using a touchpad on a laptop atm

**Comment by chrisduerr** [comment]
@smason1995 I'd recommend testing with 0.4.0 again.

**Comment by rawkode** [comment]
Still doesn't work with 0.4.0

**Comment by ranisalt** [comment]
I am not having issues with 0.4.0

**Comment by leehambley** [comment]
I'm infrequently (and it works without a restart, just by trying again now) seeing:

```
$                                                      bracketed-paste-magic:zle:47: not enough arguments for -U
[~/Videos/Winter Armada/Winter Armada Day 2]$ ffmpeg -i "2019-12-18 10-06-06.flv" -codec copy "2019-12-18 10-06-06.mp4"
ffmpeg version 4.2.1 Copyright (c) 2000-2019 the FFmpeg developers
  built with gcc 9 (GCC)
  configuration: --prefix=/usr --bindir=/usr/bin --datadir=/usr/share/ffmpeg --docdir=/usr/share/doc/ffmpeg --incdir=/usr/include/ffmpeg --libdir=/usr/lib64 --mandir=/usr/share/man --arch=x86_64 --optflags='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOU
```

for example.. this was when trying to copy the date from a filename and paste it into the last quoted section of this ffmpeg command.

**Comment by tabell** [comment]
I am seeing this issue using Sway window manager on Void Linux, and alacritty 0.4.0 (from repos). It seems to affect terminal windows after they have been open for a while, or possibly after the machine goes to sleep and wakes up. 

On a fresh terminal window, copying and pasting both within the window and to other windows, using ctrl+shift+c and ctrl+shift+v works fine. At some point it stops working for old windows, but will work on fresh ones.

**Comment by cole-h** [comment]
@tabell Your problem is more in line with https://github.com/jwilm/alacritty/issues/3050. AFAICT, this issue is talking about pasting into `alacritty` sometimes not working properly. Your problem is that sleeping and resuming the computer can kill our clipboard.

**Comment by dsbrgg** [comment]
Having this same issue here in a small project, using [copypasta](https://github.com/alacritty/copypasta).

```
alacritty 0.4.1
GNOME 3.34.2
Pop-OS 19.10
```

**Comment by noamelf** [comment]
I also have the issue:

```
alacritty 0.5.0-dev (e272572)
Ubuntu 19.10
zsh 5.7.1 (x86_64-ubuntu-linux-gnu)
```

On ctrl-shift-v I get:
```
bracketed-paste-magic:zle:47: not enough arguments for -U
```

**Comment by kchibisov** [comment]
Could someone post `env WAYLAND_DEBUG=1 alacritty` log from the repro?

**Comment by trashhalo** [comment]
> I also have the issue:
> 
> ```
> alacritty 0.5.0-dev (e272572)
> Ubuntu 19.10
> zsh 5.7.1 (x86_64-ubuntu-linux-gnu)
> ```
> 
> On ctrl-shift-v I get:
> 
> ```
> bracketed-paste-magic:zle:47: not enough arguments for -U
> ```

same issue. @kchibisov 

[term.log](https://github.com/alacritty/alacritty/files/4846993/term.log)
Steps:
1. Opened with `env WAYLAND_DEBUG=1 alacritty`
2. Paste into window see `bracketed-paste-magic:zle:47: not enough arguments for -U`
3. Close window

**Comment by kchibisov** [comment]
Could you prove the output one time, but from  `WAYLAND_DEBUG=1 alacritty --print-events > log 2>&1`?

**Comment by trashhalo** [comment]
> Could you prove the output one time, but from `WAYLAND_DEBUG=1 alacritty --print-events > log 2>&1`?

[term.log](https://github.com/alacritty/alacritty/files/4847059/term.log)

**Comment by kchibisov** [comment]
I don't see any keyboard events, were you pasting with mouse?`

**Comment by trashhalo** [comment]
Thats correct. Keyboard paste with ctrl shift v works. middle mouse paste outputs `bracketed-paste-magic:zle:47: not enough arguments for -U` into the term.

**Comment by kchibisov** [comment]
It seems like your system doesn't support primary selection at all, since the protocol isn't being advertise. Is alacritty hanging if you paste, let's say in `cat`?

**Comment by trashhalo** [comment]
> Is alacritty hanging if you paste, let's say in cat?

Im sorry, could you say more words? Im not sure what this means.

**Comment by kchibisov** [comment]
I mean that on Wayland primary selection aka middle mouse paste is using a special protocol, which is not advertise by your compositor(mutter). So we don't have primary selection at all. The normal Ctrl + Shift + V is using other protocol, that's why it works for you.

So, I'm asking you to test, whether alacritty hangs, if you run `cat` and paste with middle mouse button, since the only thing we can do here is to prevent the hang. We can't do anything about not working middle mouse, since it's 'expected' here.

**Comment by kevr** [comment]
> So, I'm asking you to test, whether alacritty hangs, if you run cat and paste with middle mouse button, since the only thing we can do here is to prevent the hang. We can't do anything about not working middle mouse, since it's 'expected' here.

Why is this closed? I believe this issue still exists.

Recently beginning to use sway and Wayland myself, I have begun to encounter this issue. What the hell do you mean that it's expected? A hang is expected?

At times, this hang forces me to completely restart the window manager (sway).

**Comment by joshuachp** [comment]
I closed as the author since this issue was 2y old, and the initial problem was solved around the 0.5.0 release (at least for me). 
Besides, Alacritty never hanged for me in neither Gnome nor Sway, so I believe the issue is somewhat different.
Wouldn't It be better to open a new Issue with a more recent log?

**Comment by kevr** [comment]
Yeah, definitely would be. Apologies; thanks, joshua.
